libraries/AP_Math/crc.cpp

```diff
@@ -304,6 +304,20 @@ uint16_t calc_crc_modbus(uint8_t *buf, uint16_t len)
     return crc;
 }
 
+// fletcher 16 implementation
+uint16_t crc_fletcher16(const uint8_t *buffer, uint32_t len) {
+    uint16_t c0 = 0;
+    uint16_t c1 = 0;
+    for (uint32_t i = 0; i < len; i++) {
+        c0 = (c0 + buffer[i]) % 255;
+        c1 = (c1 + c0) % 255;
+    }
+
+    int32_t crc = c0 - c1;
+    crc <<= 8;
+    return (c1 << 8) | c0;
+}
```

Comment:
Please see AP_ADSB_Sagetech::checksum_generate_XP(). That method is much easier to read and is computensivly trivial compared to this method. After this goes in I'll change the Sagetech driver to use the math driver.

Reply:
```suggestion
uint16_t crc_fletcher16(const uint8_t *buffer, uint32_t len) {
    uint8_t sum = 0;
    uint8_t sumFletcher = 0;
    
    for (uint32_t i = 0; i < len; i++) {
        sum += buffer[i]
        sumFletcher += sum;
    }

    return UINT16_VALUE(sum, sumFletcher);
}
```

Reply:
That suggestion doesn't result in anything close to the same value.

``` c
#include <stdlib.h>    
#include <stdio.h>    
#include <stdint.h>    
#include <string.h>    
uint16_t crc_fletcher16(const uint8_t *buffer, uint32_t len) {    
    uint16_t c0 = 0;    
    uint16_t c1 = 0;    
    for (uint32_t i = 0; i < len; i++) {    
        c0 = (c0 + buffer[i]) % 255;    
        c1 = (c1 + c0) % 255;    
    }    
    
    int32_t crc = c0 - c1;    
    crc <<= 8;    
    return (c1 << 8) | c0;    
}    
    
uint16_t tom_crc_fletcher16(const uint8_t *buffer, uint32_t len) {    
    uint8_t sum = 0;    
    uint8_t sumFletcher = 0;    
        
    for (uint32_t i = 0; i < len; i++) {    
        sum += buffer[i];    
        sumFletcher += sum;    
    }    
    return (sum << 8 ) | sumFletcher;    
}    
    
int main (int argc, char **argv) {    
  printf("0x%04x 0x%04x\n", crc_fletcher16("abcde", strlen("abcde")), tom_crc_fletcher16("abcde", strlen("abcde")));    
  return 0;                                                                                                                                
} 
```

``` bash
wickedshell@milton ~/code/tmp % gcc test.c
wickedshell@milton ~/code/tmp % ./a.out 
0xc8f0 0xefc3
```

For the record you can find test vectors [here](https://en.wikipedia.org/wiki/Fletcher%27s_checksum#Test_vectors)

Reply:
as we discussed together offline, the sagetech driver that I implemented is wrong because Sagetech is not actually a FletcherSum as they say it is. They implemented it wrong. Your version is correct. "Sagetech FlectchSum" is effectively doing mod 256 where a real FletcherSum is doing mod 255.